### PR TITLE
build(mocha-test-setup): use node16 module resolution

### PR DESCRIPTION
**Pull request by DLehenbauer**
Semi-automatic PR to switch to node16 module resolution

### CODE REVIEW

File: packages/test/mocha-test-setup/package.json [hunk 1]
```diff
@@ -75,8 +110,9 @@
 				"script": false
 			},
 			"tsc": [
-				"...",
-				"typetests:gen"
+				"typetests:gen",
+				"build:genver",
+				"^tsc"
 			]
```
Comment (jason-ha) [inline packages/test/mocha-test-setup/package.json:116]: Looks wrong. This doesn't have typetests, don't know what that is here for. In general, I think ... is preferred. The policy messages can be a little misleading. And if tsc needs build:genver, then I would expect build:esnext to as well.
@tylerbutler, why does this seem so wrong to me?

File: packages/test/mocha-test-setup/package.json [hunk 2]
```diff
@@ -11,21 +11,56 @@
...
+	},
 	"main": "dist/index.js",
-	"module": "lib/index.js",
 	"types": "dist/index.d.ts",
 	"scripts": {
 		"api": "fluid-build . --task api",
-		"api-extractor:commonjs": "api-extractor run --local",
-		"api-extractor:esnext": "api-extractor run --config ./api-extractor-esm.json",
+		"api-extractor:commonjs": "api-extractor run --config ./api-extractor-cjs.json",
+		"api-extractor:esnext": "api-extractor run --local",
```
Comment (jason-ha) [inline packages/test/mocha-test-setup/package.json:55]: We shouldn't need these as this in an internal package.
I _think_ the preferred changes for api/docs command are as shown in the test-driver-definitions PR #19917.

**Review by jason-ha — CHANGES_REQUESTED** [review]

File: packages/test/mocha-test-setup/package.json (hunk 1 above)
Comment (sonalideshpandemsft) [reply packages/test/mocha-test-setup/package.json:116]: did a small test by removing the fluidBuild config. the policy check states mocha-test-setup as a public package: `Policy violations for public package "@fluid-internal/mocha-test-setup":`

Comment (sonalideshpandemsft) [reply packages/test/mocha-test-setup/package.json]: removed the typetests:gen scripts and the `tsc` fluidBuild config

File: packages/test/mocha-test-setup/package.json [hunk 3]
```diff
@@ -11,21 +11,33 @@
...
-		"api": "fluid-build . --task api",
-		"api-extractor:commonjs": "api-extractor run --local",
-		"api-extractor:esnext": "api-extractor run --config ./api-extractor-esm.json",
 		"build": "fluid-build . --task build",
 		"build:commonjs": "fluid-build . --task commonjs",
 		"build:compile": "fluid-build . --task compile",
 		"build:compile:min": "npm run build:compile",
 		"build:docs": "fluid-build . --task api",
-		"build:esnext": "tsc --project ./tsconfig.esnext.json",
+		"build:esnext": "tsc --project ./tsconfig.json",
```
Comment (jason-ha) [inline packages/test/mocha-test-setup/package.json:38]: build:docs, I think should be `api-extractor run --local`
Then the fluidBuild entry below goes away.

**Review by jason-ha — CHANGES_REQUESTED** [review]

File: packages/test/mocha-test-setup/api-extractor-cjs.json
Comment (jason-ha) [inline packages/test/mocha-test-setup/api-extractor-cjs.json:1]: File is now unused

File: packages/test/mocha-test-setup/api-extractor.json [hunk 4]
```diff
@@ -1,4 +1,4 @@
 {
 	"$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
-	"extends": "../../../common/build/build-common/api-extractor-base.json"
+	"extends": "../../../common/build/build-common/api-extractor-base.esm.primary.json"
```
Comment (jason-ha) [inline packages/test/mocha-test-setup/api-extractor.json:3]: Like test-driver-definitions add
```json
	"dtsRollup": {
		// all APIs are internal - no rollup needed/used
		"enabled": false
	}
```

**Review by jason-ha — CHANGES_REQUESTED** [review]
There was [this discussion](https://github.com/microsoft/FluidFramework/pull/20114/files#r1523968111) between @DLehenbauer and @tylerbutler that suggests since the only export `mochaHooks` has no docs we could skip API stuff altogether here. I am also not opinionated. If keeping remove the roll-up get (comment below) otherwise that file will be gone. :)

**Comment by sonalideshpandemsft** [comment]
While removing `build:docs` and `ci:build:docs` - policy check fails:
```ts
client-release-group-root: file failed the "npm-public-package-requirements" policy: packages/test/mocha-test-setup/package.json
client-release-group-root: Policy violations for public package "@fluid-internal/mocha-test-setup":
client-release-group-root: * Missing script: "build:docs"
client-release-group-root: * Missing script: "ci:build:docs"
```

**Comment by tylerbutler** [comment]
> While removing `build:docs` and `ci:build:docs` - policy check fails:
> 
> ```ts
> 
> client-release-group-root: file failed the "npm-public-package-requirements" policy: packages/test/mocha-test-setup/package.json
> 
> client-release-group-root: Policy violations for public package "@fluid-internal/mocha-test-setup":
> 
> client-release-group-root: * Missing script: "build:docs"
> 
> client-release-group-root: * Missing script: "ci:build:docs"
> 
> ```

I like that policy! So this is a public package, which makes sense to require docs for. But maybe this is mostly our own infra and we should just add an exception to the fluidBuild config.

**Comment by tylerbutler** [comment]
Are there actually any docs?

**Comment by jason-ha** [comment]
> Are there actually any docs?

Not from code. api.md:
```ts
// @internal (undocumented)
export const mochaHooks: {
    beforeAll(): void;
    beforeEach(this: Mocha.Context): void;
    afterEach(this: Mocha.Context): void;
};

// (No @packageDocumentation comment for this package)
```
Though easy enough to keep as is now - all in place if that changes.

File: packages/test/mocha-test-setup/api-extractor-cjs.json
Comment (jason-ha) [reply packages/test/mocha-test-setup/api-extractor-cjs.json:1]: I think last change needed is to remove this unused api-extractor-cjs.json file.

**Review by jason-ha — APPROVED** [review]

**Comment by jason-ha** [comment]
> This error is part of our handler: https://github.com/microsoft/FluidFramework/blob/main/build-tools/packages/build-cli/src/library/repoPolicyCheck/npmPackages.ts#L1508. packages are not checked as per scopes (@fluid-internal, etc) I guess?

I think policy goes off of the feeds/public.txt file rather than package scope. That file indicates that the package is accessible(publicly published) even if it shouldn't be directly used. You can argue keying off either, but probably policy should kick in if either package scope says for public consumption OR is published publicly. If it is out there, we should say something about it even if it is to say don't look behind the curtain.

**Pull request merged**